File: Sources/QuickObjectiveC/QuickSpec.m
#import "QuickSpec.h"
#import "QuickConfiguration.h"

#if __has_include("Quick-Swift.h")
#import "Quick-Swift.h"
#else
#import <Quick/Quick-Swift.h>
#endif

static QuickSpec *currentSpec = nil;

@interface QuickSpec ()
@property (nonatomic, strong) Example *example;
@end

@implementation QuickSpec

#pragma mark - XCTestCase Overrides

/**
 Invocations for each test method in the test case. QuickSpec overrides this method to define a
 new method for each example defined in +[QuickSpec spec].

 @return An array of invocations that execute the newly defined example methods.
 */
+ (NSArray *)testInvocations {
    // Xcode 13.3 hack, see this issue for more info: https://github.com/Quick/Quick/issues/1123
    // In case of fix in later versions next line can be removed
    [[QuickTestObservation sharedInstance] buildAllExamplesIfNeeded];

    NSArray<ExampleWrapper *> *examples = [[World sharedWorld] examplesForSpecClass:[self class]];
    NSMutableArray *invocations = [NSMutableArray arrayWithCapacity:[examples count]];
    
    NSMutableSet<NSString*> *selectorNames = [NSMutableSet set];
    
    for (ExampleWrapper *exampleWrapper in examples) {
        SEL selector = [self addInstanceMethodForExample:exampleWrapper.example runFullTest:exampleWrapper.runFullTest classSelectorNames:selectorNames];

        NSMethodSignature *signature = [self instanceMethodSignatureForSelector:selector];
        NSInvocation *invocation = [NSInvocation invocationWithMethodSignature:signature];
        invocation.selector = selector;

        [invocations addObject:invocation];
    }

    return invocations;
}

/**
 This method is used as a hook for injecting test methods into the
 Objective-C runtime on individual test runs.
 
 When `xctest` runs a test on a single method, it does not call
 `defaultTestSuite` on the test class but rather calls
 `instancesRespondToSelector:` to build its own suite.
 
 In normal conditions, Quick uses the implicit call to `defaultTestSuite`
 to both generate examples and inject them as methods by way of
 `testInvocations`.  Under single test conditions, there's no implicit
 call to `defaultTestSuite` so we make it explicitly here.
 */
+ (BOOL)instancesRespondToSelector:(SEL)aSelector {
    [self defaultTestSuite];
    return [super instancesRespondToSelector:aSelector];
}

#pragma mark - Public Interface

+ (void)spec { }

+ (QuickSpec*) current {
    return currentSpec;
}

#pragma mark - Internal Methods

/**
 Runs the `spec` method and builds the examples for this class.

 It's safe to call this method multiple times. If the examples for the class have been built, invocation
 of this method has no effect.
 */
+ (void)buildExamplesIfNeeded {
    [QuickConfiguration class];
    World *world = [World sharedWorld];

    if ([world isRootExampleGroupInitializedForSpecClass:[self class]]) {
        // The examples for this subclass have been already built. Skipping.
        return;
    }

    ExampleGroup *rootExampleGroup = [world rootExampleGroupForSpecClass:[self class]];
    [world performWithCurrentExampleGroup:rootExampleGroup closure:^{
        @try {
            [self spec];
        }
        @catch (NSException *exception) {
            [NSException raise:NSInternalInconsistencyException
                        format:@"An exception occurred when building Quick's example groups.\n"
             @"Some possible reasons this might happen include:\n\n"
             @"- An 'expect(...).to' expectation was evaluated outside of "
             @"an 'it', 'context', or 'describe' block\n"
             @"- 'sharedExamples' was called twice with the same name\n"
             @"- 'itBehavesLike' was called with a name that is not registered as a shared example\n\n"
             @"Here's the original exception: '%@', reason: '%@', userInfo: '%@'",
             exception.name, exception.reason, exception.userInfo];
        }
    }];
}

/**
 QuickSpec uses this method to dynamically define a new instance method for the
 given example. The instance method runs the example, catching any exceptions.
 The exceptions are then reported as test failures.

 In order to report the correct file and line number, examples must raise exceptions
 containing following keys in their userInfo:

 - "SenTestFilenameKey": A String representing the file name
 - "SenTestLineNumberKey": An Int representing the line number

 These keys used to be used by SenTestingKit, and are still used by some testing tools
 in the wild. See: https://github.com/Quick/Quick/pull/41

 @return The selector of the newly defined instance method.
 */
+ (SEL)addInstanceMethodForExample:(Example *)example runFullTest:(BOOL)runFullTest classSelectorNames:(NSMutableSet<NSString*> *)selectorNames {
    IMP implementation = imp_implementationWithBlock(^(QuickSpec *self){
        self.example = example;
        currentSpec = self;
        if (runFullTest) {
            [example run];
        } else {
            [example runSkippedTest];
        }
        currentSpec = nil;
    });

    const char *types = [[NSString stringWithFormat:@"%s%s%s", @encode(void), @encode(id), @encode(SEL)] UTF8String];

    NSString *selectorName = [TestSelectorNameProvider testSelectorNameFor:example classSelectorNames:selectorNames];

    [selectorNames addObject:selectorName];

    SEL selector = NSSelectorFromString(selectorName);
    class_addMethod(self, selector, implementation, types);

    return selector;
}

/**
 This method is used to record failures, whether they represent example
 expectations that were not met, or exceptions raised during test setup
 and teardown. By default, the failure will be reported as an
 XCTest failure, and the example will be highlighted in Xcode.
 */
- (void)recordIssue:(XCTIssue *)issue {
    if (self != [QuickSpec current]) {
        [[QuickSpec current] recordIssue:issue];
        return;
    }

    if (self.example.isSharedExample) {
        XCTSourceCodeLocation *location = [[XCTSourceCodeLocation alloc] initWithFilePath:self.example.callsite.file
                                                                               lineNumber:self.example.callsite.line];
        XCTSourceCodeContext *sourceCodeContext = [[XCTSourceCodeContext alloc] initWithLocation:location];
        XCTIssue *newIssue = [[XCTIssue alloc] initWithType:issue.type
                                         compactDescription:issue.compactDescription
                                        detailedDescription:issue.detailedDescription
                                          sourceCodeContext:sourceCodeContext
                                            associatedError:issue.associatedError
                                                attachments:issue.attachments];
        [super recordIssue:newIssue];
    } else {
        [super recordIssue:issue];
    }
}

@end

#pragma mark - Test Observation

__attribute__((constructor))
static void registerQuickTestObservation(void) {
    [[XCTestObservationCenter sharedTestObservationCenter] addTestObserver:[QuickTestObservation sharedInstance]];
}

File: Sources/Quick/QuickSpec.swift
import XCTest

// NOTE: This file is not intended to be included in the Xcode project or CocoaPods.
//       It is picked up by the Swift Package Manager during its build process.

#if SWIFT_PACKAGE

#if canImport(QuickObjCRuntime)
import QuickObjCRuntime

public typealias QuickSpecBase = _QuickSpecBase
#else
public typealias QuickSpecBase = XCTestCase
#endif

open class QuickSpec: QuickSpecBase {
    /// Returns the currently executing spec. Use in specs that require XCTestCase
    /// methods, e.g. expectation(description:).
    ///
    /// If you're using `beforeSuite`/`afterSuite`, you should consider the ``currentSpec()`` helper.
    public private(set) static var current: QuickSpec!

    private var example: Example? {
        didSet {
            QuickSpec.current = self
        }
    }

    @MainActor
    open class func spec() {}

#if canImport(Darwin)
    /// This method is used as a hook for the following two purposes
    ///
    /// 1. Performing all configurations
    /// 2. Gathering examples for each spec classes
    ///
    /// On Linux, those are done in `LinuxMain.swift` and `Quick.QCKMain`. But
    /// SwiftPM on macOS does not have the mechanism (test cases are automatically
    /// discovered powered by Objective-C runtime), so we needed the alternative
    /// way.
    override open class var defaultTestSuite: XCTestSuite {
        QuickConfiguration.configureSubclassesIfNeeded(world: World.sharedWorld)

        // Let's gather examples for each spec classes. This has the same effect
        // as listing spec classes in `LinuxMain.swift` on Linux.
        gatherExamplesIfNeeded()

        return super.defaultTestSuite
    }

    /// This method is used as a hook for injecting test methods into the
    /// Objective-C runtime on individual test runs.
    ///
    /// When `xctest` runs a test on a single method, it does not call
    /// `defaultTestSuite` on the test class but rather calls
    /// `instancesRespondToSelector:` to build its own suite.
    ///
    /// In normal conditions, Quick uses the implicit call to `defaultTestSuite`
    /// to both generate examples and inject them as methods by way of
    /// `testInvocations`.  Under single test conditions, there's no implicit
    /// call to `defaultTestSuite` so we make it explicitly here.
    open override class func instancesRespond(to aSelector: Selector!) -> Bool {
        _ = self.defaultTestSuite
        return super.instancesRespond(to: aSelector)
    }

    override open class func _qck_testMethodSelectors() -> [String] {
        let examples = World.sharedWorld.examples(forSpecClass: self)

        var selectorNames = Set<String>()
        return examples.map { (test: ExampleWrapper) in
            let selector = addInstanceMethod(for: test.example, runFullTest: test.runFullTest, classSelectorNames: &selectorNames)
            return NSStringFromSelector(selector)
        }
    }

    private static func addInstanceMethod(for example: Example, runFullTest: Bool, classSelectorNames selectorNames: inout Set<String>) -> Selector {
        let block: @convention(block) @MainActor (QuickSpec) -> Void = { spec in
            spec.example = example
            if runFullTest {
                example.run()
            } else {
                example.runSkippedTest()
            }
            QuickSpec.current = nil
        }
        let implementation = imp_implementationWithBlock(block as Any)

        let selectorName = TestSelectorNameProvider.testSelectorName(for: example, classSelectorNames: selectorNames)

        selectorNames.insert(selectorName)

        let selector = NSSelectorFromString(selectorName)
        class_addMethod(self, selector, implementation, "v@:")

        return selector
    }
#endif // canImport(Darwin)

#if !canImport(Darwin)
    public required init() {
        super.init(name: "", testClosure: { _ in })
    }

    public required init(name: String, testClosure: @escaping (XCTestCase) throws -> Swift.Void) {
        super.init(name: name, testClosure: testClosure)
    }

    public class var allTests: [(String, (QuickSpec) -> () throws -> Void)] {
        gatherExamplesIfNeeded()

        let exampleWrappers = World.sharedWorld.examples(forSpecClass: self)
        let result = exampleWrappers.map { wrapper -> (String, (QuickSpec) -> () throws -> Void) in
            return (wrapper.example.name, { spec in
                let example = wrapper.example
                return { @MainActor in
                    spec.example = example
                    if wrapper.runFullTest {
                        example.run()
                    } else {
                        example.runSkippedTest()
                    }
                    QuickSpec.current = nil
                }
            })
        }
        return result
    }
#endif

    internal static func gatherExamplesIfNeeded() {
        let world = World.sharedWorld
        let rootExampleGroup = world.rootExampleGroup(forSpecClass: self)
        guard rootExampleGroup.examples.isEmpty else {
            return
        }

        world.performWithCurrentExampleGroup(rootExampleGroup) {
            self.spec()
        }
    }

    // MARK: Delegation to `QuickSpec.current`.

    override public func recordFailure(
        withDescription description: String,
        inFile filePath: String,
        atLine lineNumber: Int,
        expected: Bool
    ) {
        guard self === Self.current else {
            Self.current.recordFailure(
                withDescription: description,
                inFile: filePath,
                atLine: lineNumber,
                expected: expected
            )
            return
        }

        super.recordFailure(
            withDescription: description,
            inFile: filePath,
            atLine: lineNumber,
            expected: expected
        )
    }
}

#endif
